### PR TITLE
Implement report.rs component

### PR DESCRIPTION
**Pull request by ranya-cfa**
Work in progress

### CODE REVIEW

**Comment by ranya-cfa** [comment]
"".to_string() giving cargo clippy error

File: src/report.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+impl ContextReportExt for Context {
+    fn generate_filename(&mut self, short_name: &str) -> PathBuf {
+        let data_container = self.get_data_container_mut::<ReportPlugin>();
+        let prefix = data_container.config.file_prefix.clone();
+        let directory = data_container.config.directory.clone();
+        let short_name = short_name.to_string();
+        let basename = format!("{prefix}{short_name}");
+        let filename = Path::new(&directory).join(basename).with_extension("csv");
+
+        println!("Generated filename: {}", filename.display());
```
Comment (k88hudson-cfa) [inline src/report.rs:81]: Remove this println

File: src/report.rs [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+pub trait ContextReportExt {
+    fn generate_filename(&mut self, short_name: &str) -> PathBuf;
+    fn add_report<T: Report + 'static>(&mut self, short_name: &str);
+    fn send_report<T: Report>(&mut self, report: T);
+    fn init_reports(&mut self, file_prefix: Option<&str>, directory: Option<&str>);
+}
+
+impl ContextReportExt for Context {
+    fn generate_filename(&mut self, short_name: &str) -> PathBuf {
```
Comment (k88hudson-cfa) [inline src/report.rs:73]: Rather than making this part of the trait extension, can you make this a private helper function?

File: src/report.rs [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+            .get_mut(&report.type_id())
+        {
+            report.serialize(writer);
+            writer.flush().expect("Failed to flush writer"); // Needed for tests to work
+        } else {
+            panic!("No writer found for the report type");
+        }
+    }
+
+    fn init_reports(&mut self, file_prefix: Option<&str>, directory: Option<&str>) {
```
Comment (k88hudson-cfa) [inline src/report.rs:111]: let's call this `set_report_options`

**Review by k88hudson-cfa — CHANGES_REQUESTED** [review]

File: src/report.rs [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+    }
+
+    #[test]
+    fn test_csv_serialization() -> Result<(), Box<dyn Error>> {
+        let mut writer = Writer::from_writer(vec![]);
+
+        let data = vec![
+            TestStruct {
+                id: 1,
+                name: "Test1".to_string(),
```
Comment (k88hudson-cfa) [inline src/report.rs:348]: Can you test something with a Comma, like "Test,1"

File: src/report.rs [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+            TestStruct {
+                id: 2,
+                name: "Test2".to_string(),
+                active: false,
+                category: TestEnum::Category2,
+            },
+        ];
+
+        for record in data {
+            writer.serialize(record)?;
```
Comment (k88hudson-cfa) [inline src/report.rs:361]: You should be testing the output of the library itself, not constructing a writer

File: src/report.rs [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+
+        // Convert the in-memory writer to a string
+        let csv_data = String::from_utf8(writer.into_inner()?)?;
+
+        // Verify content
+        let expected_csv =
+            "id,name,active,category\n1,Test1,true,Category1\n2,Test2,false,Category2\n";
+        assert_eq!(csv_data, expected_csv);
+
+        Ok(())
```
Comment (k88hudson-cfa) [inline src/report.rs:374]: Is this intended?

**Review by k88hudson-cfa — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 7]
```diff
@@ -8,3 +8,7 @@ license = "Apache-2.0"
 homepage = "https://github.com/CDCgov/ixa"
 
 [dependencies]
+csv = "1.3" # or the version you're using
+serde = { version = "1.0", features = ["derive"] }
+serde_derive = "1.0"
+tempfile = "3.3"
```
Comment (k88hudson-cfa) [inline Cargo.toml:14]: Should be dev dependencies

File: src/report.rs [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+        let mut writer_cell = data_container.file_writers.try_borrow_mut().unwrap();
+        let writer = writer_cell
+            .get_mut(&report.type_id())
+            .expect("No writer found for the report type");
+        report.serialize(writer);
+        writer.flush().expect("Failed to flush writer");
+    }
+
+    /// This is the builder for the report options. It must be called on context
+    /// The options can be assigned to the config that is returned from this function
```
Comment (k88hudson-cfa) [inline src/report.rs:139]:
```suggestion
    /// Returns a `ConfigReportOptions` object which has setter methods for report configuration
```

File: src/report.rs [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+
+        let file = File::create(path).expect("Couldn't create file");
+        let writer = Writer::from_writer(file);
+        let mut file_writer = data_container.file_writers.borrow_mut();
+        file_writer.insert(TypeId::of::<T>(), writer);
+    }
+
+    /// Anytime something happens (ie. a person gets infected) in the model that
+    /// you want to record, you can use `send_report` to write a new row to a report.
+    /// Note that you will need to call the method with the appropriate report type
```
Comment (k88hudson-cfa) [inline src/report.rs:124]:
```suggestion
    /// Write a new row to the appropriate report file
```

File: src/report.rs [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+
+pub trait ContextReportExt {
+    fn add_report<T: Report + 'static>(&mut self, short_name: &str);
+    fn send_report<T: Report>(&self, report: T);
+    fn report_options(&mut self) -> &mut ConfigReportOptions;
+}
+
+impl ContextReportExt for Context {
+    /// Call `add_report` with each report type, passing the name of the report type.
+    /// The report type name is used for file naming to distinguish what data each
```
Comment (k88hudson-cfa) [inline src/report.rs:109]:
```suggestion
    /// The `short_name` is used for file naming to distinguish what data each
```

**Review by k88hudson-cfa — APPROVED** [review]
LGTM

**Pull request merged**